File: test2/BIS_pass.R
# 1 -----------------------------------------------------------------------
cat("\014")
rm(list=ls())
library(tidyverse)
library(readxl)
url <- "https://dsp.twse.com.tw/public/static/downloads/brokerDepartment/肆、測驗合格人員名單_20250103153040.xlsx"
dest <- "Exam_Pass_List.xlsx"
download.file(url, destfile = dest, mode = "wb")
da <- read_excel(dest, col_names = c("券商代號", "券商名稱", "姓名"))
if (file.exists(dest)) unlink(dest)
tmp        <- da[, 1, drop=TRUE]
header_idx <- which(tmp == "券商代號")
year_idx   <- header_idx - 1
year       <- as.integer(str_extract(tmp[year_idx], "\\d+(?=年)"))
idx_start  <- header_idx + 1
idx_end    <- c(year_idx[-1] - 1, nrow(da))

df <- tibble(
  券商代號 = character(),
  券商名稱 = character(),
  姓名 = character()
)
for (i in seq_along(year)) {
  df_tmp <- da %>% slice(idx_start[i]:idx_end[i]) %>% drop_na(姓名) %>% mutate(year = year[i])
  df <- bind_rows(df, df_tmp)
}
df
df %>% filter(str_starts(姓名, ""), str_ends(姓名, ""))

df %>% 
  filter(str_starts(券商名稱, "")) %>%
  filter(str_starts(姓名, ""), str_ends(姓名, ""))

File: test2/BIS_pass.py
import pandas as pd
import requests
import os

url = "https://dsp.twse.com.tw/public/static/downloads/brokerDepartment/肆、測驗合格人員名單_20250103153040.xlsx"
file_name = "Exam_Pass_List.xlsx"

r = requests.get(url)
with open(file_name, "wb") as f:
    f.write(r.content)

da = pd.read_excel(file_name, header=None, names=["券商代號", "券商名稱", "姓名"])
os.remove(file_name)
tmp = da["券商代號"]

header_idx = tmp[tmp == "券商代號"].index
year_idx = header_idx - 1

year = (
    tmp.loc[year_idx]
    .str.extract(r"(\d+)(?=年)")[0]
    .astype(int)
    .tolist()
)

idx_start = header_idx + 1
idx_end = list(year_idx[1:] - 1) + [len(da) - 1]
from pprint import pprint
pprint(list(zip(idx_start, idx_end, year)))
dfs = [
    da.loc[s:e]
      .dropna(subset=["姓名"])
      .assign(year=y)
    for s, e, y in zip(idx_start, idx_end, year)
]

df = pd.concat(dfs, ignore_index=True)
print(df)
print(df[
    df["姓名"].str.startswith("") &
    df["姓名"].str.endswith("")
])

print(df[
    df["券商名稱"].str.startswith("") &
    df["姓名"].str.startswith("") &
    df["姓名"].str.endswith("")
])

print(df[
    df["券商名稱"].str.startswith("") &
    df["姓名"].str.startswith("") &
    df["姓名"].str.endswith("")
])
